specs/CameraCaptureUI/CameraCaptureUI.md

```diff
@@ -0,0 +1,387 @@
+
+This spec details the API of a new `CameraCaptureUI` in WinAppSDK,
+enabling desktop applications such as WinUI3 to seamlessly integrate native camera capture 
+functionality across diverse Windows platforms.
+
+
+
+# Background
+
+The current OS 
+[CameraCaptureUI](https://docs.microsoft.com/uwp/api/Windows.Media.Capture.CameraCaptureUI),
+primarily designed for UWP applications,
+presents several challenges when used in desktop environments. 
+Key issues include the dependency on `CoreWindow`, 
+which are not compatible with desktop app frameworks and the lack of support for `IInitializeWithWindow`. 
+As a result, developers are forced to resort to workarounds involving interop interfaces and 
+additional code to integrate camera capture functionality into their WinUI 3 desktop applications.
+
+In response to this feedback, this `CameraCaptureUI` API for WinAppSDK aims to provide 
+a streamlined solution that meets the expectations of our developer community. 
+The goal is to offer seamless integration of camera capture capabilities with 
+full feature parity across all Windows platforms supported by WinAppSDK. 
+This new API eliminates the need for cumbersome workarounds and 
+simplifies the process of incorporating camera functionality into desktop applications.
+
+
+This new WinAppSDK version of the `CameraCaptureUI` is essentially the same as the existing public
+[CameraCaptureUI](https://learn.microsoft.com/en-us/uwp/api/windows.media.capture.cameracaptureui?view=winrt-22621) 
+API, with a minor adjustment to the CameraCapture constructor to require a
+[WindowId](https://docs.microsoft.com/windows/windows-app-sdk/api/winrt/Microsoft.UI.WindowId).
+
+
+
+
+
+
+# Conceptual pages
+
+## Capture a photo (C++)
+
+A C++ example that tests whether a photo was successfully taken using the CaptureFileAsync method.
+
+```c++
+    // Retrieve the window handle (HWND) of the current WinUI 3 window.
+    auto windowNative{ this->m_inner.as<::IWindowNative>() };
+    HWND hWnd{ 0 };
+    windowNative->get_WindowHandle(&hWnd);
+
+    // Get the WindowId for the window
+    Microsoft::UI::WindowId windowId = Microsoft::UI::GetWindowIdFromWindow(hWnd);
+
+    // Initialize CameraCaptureUI with a window handle
+    winrt::Microsoft::Windows::Media::Capture::CameraCaptureUI cameraUI(windowId);
+
+    // Configure Photo Settings
+    cameraUI.PhotoSettings().Format(CameraCaptureUIPhotoFormat::Jpeg);
+    cameraUI.PhotoSettings().AllowCropping(false);
+
+    // Capture a photo asynchronously
+    auto photo = co_await cameraUI.CaptureFileAsync(CameraCaptureUIMode::Photo);
+
+    // Assert the result
+    if (photo != nullptr)
+    {
+        // Log success if photo capture was successful
+        Log::Comment(L"Photo capture was successful.");
+    }
+    else
+    {
+        // Log error if photo capture failed or was canceled
+        Log::Error(L"Photo capture failed or was canceled.");
```

Comment:
If it failed then the IAsyncOperation should have thrown.  Cancel should have thrown a canceled exception.

Reply:
Does the cancel case also return null? 
For UWP apps using the UWP CameraCaptureUI API, it doesnt throw an exception and returns file as null in such cases.![image](https://github.com/user-attachments/assets/78cd5c1f-405a-4b48-8b99-6ad8cc22165b)

However, for WinUI apps using the WinAppSDK version of the CameraCaptureUI API, it returns null, along with throwing an exception E_INVALIDARG as well. Is this behavior okay as this behavior differs from UWP version.
![image](https://github.com/user-attachments/assets/47927e3a-b668-4e0c-9979-a34975d21053)

Reply:
@dmachaj **"If it failed then the IAsyncOperation should have thrown. Cancel should have thrown a canceled exception."**

For the WinAppSDK version of the CameraCaptureUI API, it returns null and throws an E_INVALIDARG exception, which seems like to align with the behavior you mentioned and also sounds like a bug in my code, ideally my implementation should be handling this correct?

However, the key question is whether we should throw an exception for a cancel operation. In UWP, cancellation is handled by returning null, which is a straightforward way to manage this scenario.

Reply:
cppwinrt has a specific `winrt::hresult_canceled` exception type specifically for this case.  See https://github.com/microsoft/cppwinrt/blob/0deecf1ea4fcc6acd10602e2340e243fcdcf4f47/strings/base_error.h#L329.  That corresponds to `HRESULT_FROM_WIN32(ERROR_CANCELLED)`.

Reply:
That said, exceptions are supposed to be for exceptional circumstances.  The user deciding not to take a photo probably shouldn't count as exceptional.  As a result, S_OK+nullptr is probably the right approach.  (Other errors should still throw).

Reply:
Agreed: user canceling is expected and should be S_OK+nullptr. An error case like the Camera app crashing probably should too.

The guideline is that apps using this API shouldn't be expected or recommended to have a try/catch; if an exception is thrown to the app it's a sign of an app bug.

Reply:
Saharsh and I was discussing about various cases and current API implementation.

**Success cases:**
* User captures a file [S_OK+File+No Exception]
* User presses the cancel button [S_OK+nullptr+No Exception]
**Failure cases:**
* If Camera app is being used already [Exception APP_NOT_AVAILABLE]
* If Camera app URI launch mapping unavailable [Exception PROTOCOL_NOT_AVAILABLE]
* Any other unknown error [Exception GENERAL_FAILURE]

Saharsh would look into more how the UWP API is implemented considering the above success and failure cases. 

In general, what is the recommendation to handle failure cases? is it recommended to throw an exception for failure cases (like mentioned above) or should some other approaches. If other approaches, I think following:
1. The return value of CaptureFileAsync API would be some other interface/class/structure than IAsyncOperation< StorageFile >. The disadvantage is API will be more complicated and non-uniform across UWP & WASDK.
2. Have separate APIs implemented to get the status of the error. This does not look like a clean way.

Reply:
Hey @MikeHillberg @dmachaj 
I've reviewed the behavior of the UWP API across these scenarios to understand how errors are being managed.

Here are the observations:

1) Successful Capture: Returns a storage object with S_OK.
2) Cancelled Capture: Returns null with S_OK. (No exceptions).
3) App Uninstalled: Displays a dialog box (Search for app in store) Yes/No? clicking on No returns null with S_OK.
4) Incorrect URI Mapping: Returns null with S_OK (No exceptions).

It seems the UWP API handles all these cases gracefully without throwing exceptions. Our goal should be to maintain consistency with this existing behavior.